modules/ml/src/main/java/org/apache/ignite/ml/naivebayes/bernoulli/BernoulliNaiveBayesTrainer.java

```diff
@@ -0,0 +1,181 @@
+
+    /**
+     * Trains model based on the specified data.
+     *
+     * @param datasetBuilder Dataset builder.
+     * @param featureExtractor Feature extractor.
+     * @param lbExtractor Label extractor.
+     * @return Model.
+     */
+    @Override public <K, V> BernoulliNaiveBayesModel fit(DatasetBuilder<K, V> datasetBuilder,
+        IgniteBiFunction<K, V, Vector> featureExtractor, IgniteBiFunction<K, V, Double> lbExtractor) {
+        return updateModel(null, datasetBuilder, featureExtractor, lbExtractor);
+    }
+
+    /** {@inheritDoc} */
+    @Override protected boolean checkState(BernoulliNaiveBayesModel mdl) {
+        return true;
+    }
+
+    /** {@inheritDoc} */
+    @Override protected <K, V> BernoulliNaiveBayesModel updateModel(BernoulliNaiveBayesModel mdl,
+        DatasetBuilder<K, V> datasetBuilder, IgniteBiFunction<K, V, Vector> featureExtractor,
+        IgniteBiFunction<K, V, Double> lbExtractor) {
+
+        try (Dataset<EmptyContext, BernoulliNaiveBayesSumsHolder> dataset = datasetBuilder.build(
+            (upstream, upstreamSize) -> new EmptyContext(),
+            (upstream, upstreamSize, ctx) -> {
+                BernoulliNaiveBayesSumsHolder res = new BernoulliNaiveBayesSumsHolder();
+                while (upstream.hasNext()) {
+                    UpstreamEntry<K, V> entity = upstream.next();
+
+                    Vector features = featureExtractor.apply(entity.getKey(), entity.getValue());
+                    Double label = lbExtractor.apply(entity.getKey(), entity.getValue());
+
+                    long[] onesCount;
+
+                    if (!res.onesCountPerLbl.containsKey(label)) {
+                        onesCount = new long[features.size()];
+                        Arrays.fill(onesCount, 0L);
+                        res.onesCountPerLbl.put(label, onesCount);
+                    }
+                    if (!res.featureCountersPerLbl.containsKey(label)) {
+                        res.featureCountersPerLbl.put(label, 0);
+                    }
+                    res.featureCountersPerLbl.put(label, res.featureCountersPerLbl.get(label) + 1);
+
+                    onesCount = res.onesCountPerLbl.get(label);
+                    for (int j = 0; j < features.size(); j++) {
+                        double x = features.get(j);
+                        if (x >= binarizeThreshold)
+                            ++onesCount[j];
+                    }
+                }
+                return res;
+            })) {
+            BernoulliNaiveBayesSumsHolder sumsHolder = dataset.compute(t -> t, (a, b) -> {
+                if (a == null)
+                    return b == null ? new BernoulliNaiveBayesSumsHolder() : b;
+                if (b == null)
+                    return a;
+                return a.merge(b);
+            });
+            if (mdl != null && mdl.getSumsHolder() != null) {
+                assert mdl.getBinarizeThreshold() == binarizeThreshold;
```

Comment:
Maybe this check can be moved upper to checkState method and in this place we should check that  feature vector sizes are same

Reply:
I added a method which checks that vector sizes are same.